phases/19-capstone-projects/52-experiment-runner/code/main.py

```
@@ -0,0 +1,300 @@
+    def run(self, spec: ExperimentSpec) -> ExperimentResult:
+        with tempfile.TemporaryDirectory(prefix="exp_") as workdir:
+            config_path = os.path.join(workdir, "config.json")
+            merged_config = dict(spec.config)
+            merged_config["__seed"] = spec.seed
+            with open(config_path, "w", encoding="utf-8") as fh:
+                json.dump(merged_config, fh)
+            return self._run_subprocess(spec, config_path)
+
+    def _run_subprocess(self, spec: ExperimentSpec, config_path: str) -> ExperimentResult:
+        start = time.perf_counter()
+        proc = subprocess.Popen(
```

Ruff (0.15.14)

[error] `subprocess` call: check for execution of untrusted input

(S603)